### DISCUSSION TITLE
How to do analytics for editor changes?

### DISCUSSION THREAD

**Discussion by strdr4605** [facebook/lexical#3151] [Q&A] [answered]
I am trying to add analytics on how users use the lexical editor. I would like to monitor text format changes (bold, italic, ...).
h1 insert, custom node insert, markdown shortcuts, keyboard shortcuts.

I found https://lexical.dev/docs/concepts/listeners#registerupdatelistener, but this listener does not should the diff.

I also thought to use the following:
```js
editor.registerCommand(FORMAT_TEXT_COMMAND, (payload) => analytics.track(...), COMMAND_PRIORITY_LOW);
```
but I should add a register with this callback for every Lexical Command.

Maybe there should be a listener for every command executed?
What is a good way to track these changes?

**Comment by strdr4605** [comment, marked as answer]
@thegreatercurve

Why there are commands for list insertion `INSERT_UNORDERED_LIST_COMMAND`, ...
but no commands for inserting headings, or paragraphs?  
Maybe it makes sense to also add these commands and use them instead of `$createHeadingNode(heading)`?

**Reply by thegreatercurve** [reply]
Simply because we've not needed to add commands for core node creations. At the moment, there's only one way of creating core nodes (though this may changes #1262), so you can call the creation methods directly. For more complex functionality, such as lists, which we've abstracted out into their own packages, developers can listen for the provided commands though themselves and create their own list creation logic, if they so wish.

**Reply by thegreatercurve** [reply]
> I found https://lexical.dev/docs/concepts/listeners#registerupdatelistener, but this listener does not should the diff.

~On this point, you can store the previous editor state in a ref outside of the callback and then diff the editor states to see what's been added or removed.~

EDIT: You can grab the previous editor state in the callback directly: 
```js
editor.registerUpdateListener(({editorState, prevEditorState}) => {
    diffAndLogEvent(editorState, previousEditorState)
}));

**Reply by trueadm** [reply]
update listeners get passed `prevEditorState` already, so you can use that to do the diff.
